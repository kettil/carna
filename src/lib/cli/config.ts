import { join } from 'path';
import { isObject } from '@kettil/tool-lib';
import access from '../cmd/access';
import readFile from '../cmd/readFile';

const getDataRecursive = (data: unknown, keys: string[]): unknown => {
  const key = keys.shift();

  if (key === undefined) {
    return data;
  }

  return isObject(data) ? getDataRecursive(data[key], keys) : undefined;
};

const getConfig = async (cwd: string, keyPath?: string): Promise<unknown> => {
  const configPath = join(cwd, '.carnarc.json');

  const hasConfigFile = await access(configPath, 'readable');

  if (!hasConfigFile) {
    // carna config file is not readable
    return undefined;
  }

  const data = await readFile(configPath, true);

  return typeof keyPath === 'string' ? getDataRecursive(data, keyPath.split('.')) : data;
};

export default getConfig;
