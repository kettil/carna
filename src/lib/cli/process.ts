const getStdin = (): NodeJS.ReadStream => process.stdin;

export { getStdin };
