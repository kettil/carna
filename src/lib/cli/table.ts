const defaultBorder = {
  t: '─',
  tl: '┌',
  tb: '┬',
  tr: '┐',

  l: '│',
  lr: '├',

  c: '┼',
  ch: '─',
  cv: '│',

  r: '│',
  rl: '┤',

  b: '─',
  bl: '└',
  bt: '┴',
  br: '┘',
};

type Border = typeof defaultBorder;
type TableBorder = (columnLengths: number[], line: string, first: string, cross: string, last: string) => string;
type TableRow = (columnLengths: number[], row: string[], separator: string, first: string, last: string) => string;

const cleanCliValue = (v: string): string => v.replace(/(\u{1B})?\[\d{1,2}m/gu, '');

const getMaxColumnLengths = (maxLengths: number[], row: string[]): number[] => {
  if (maxLengths.length > 0 && maxLengths.length !== row.length) {
    throw new Error('The table rows are of different lengths');
  }

  return row
    .map((value) => cleanCliValue(value).length)
    .map((value, index) => (typeof maxLengths[index] === 'number' ? Math.max(maxLengths[index], value) : value));
};

const cellFill = (cell: string, length: number): string => cell + ' '.repeat(length - cleanCliValue(cell).length);

const tableBorder: TableBorder = (columnLengths, line, first, cross, last) =>
  first + line + columnLengths.map((l) => line.repeat(l)).join(line + cross + line) + line + last;

const tableRow: TableRow = (columnLengths, row, separator, first, last) =>
  `${first} ${columnLengths.map((length, i) => cellFill(row[i], length)).join(` ${separator} `)} ${last}`;

const table = (rows: string[][], border: Border = defaultBorder): string => {
  const { b, bl, br, bt, c, ch, cv, l, lr, r, rl, t, tb, tl, tr } = border;
  const columnLengths = rows.reduce(getMaxColumnLengths, []);
  const separator = tableBorder(columnLengths, ch, lr, c, rl);

  return [
    tableBorder(columnLengths, t, tl, tb, tr),
    rows.map((row) => tableRow(columnLengths, row, cv, l, r)).join(`\n${separator}\n`),
    tableBorder(columnLengths, b, bl, bt, br),
  ].join('\n');
};

export { table };
