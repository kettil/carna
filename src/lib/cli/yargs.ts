import { CommandModule, Arguments, Argv } from 'yargs';
import ExecutableError from '../errors/executableError';
import { PropsGlobal } from '../types';

export type CommandModuleCommand = string;

export type CommandModuleDescribe = NonNullable<CommandModule<PropsGlobal>['describe']>;

export type CommandModuleBuilder<Props = Record<string, unknown>> = (
  yargs: Argv<PropsGlobal>,
) => Argv<PropsGlobal & Props>;

export type CommandModuleHandler<Props = Record<string, unknown>> = (args: Arguments<PropsGlobal & Props>) => void;

export const builderDefault = <Props>(cmd: string, fn: CommandModuleBuilder<Props>): CommandModuleBuilder<Props> => (
  yargs,
) => fn(yargs).usage(`Usage: $0 ${cmd} [options]`).help().version(false)
  .showHelpOnFail(false);

export const errorHandler = (argv: PropsGlobal, error: unknown): void => {
  if (error instanceof ExecutableError) {
    argv.log.log(error.stdout);

    /* eslint-disable-next-line node/no-process-exit, unicorn/no-process-exit */
    process.exit(1);
  }

  throw error;
};
