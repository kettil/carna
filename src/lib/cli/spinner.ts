import { constants } from 'os';
import { dummy } from '@kettil/tool-lib';
import ora from 'ora';
import { exit } from '../cmd/exit';
import { SpinnerWarnError } from '../errors/spinnerWarnError';
import { SpawnKillHandler } from '../utils/createSpawnKillHandler';
import { getStdin } from './process';

type StopTypes = 'fail' | 'info' | 'stop' | 'succeed' | 'warn';

const instance = ora({
  color: 'white',
  spinner: 'dots',
});

const isSpinning = (): boolean => instance.isSpinning;

const start = (text: string) => {
  instance.start(text);
};

const stop = (key: StopTypes, text?: string) => {
  if (isSpinning()) {
    instance[key](text);
  }
};

const succeed = <T>(p: T) => {
  stop('succeed');

  return p;
};

const fail = (error: unknown) => {
  if (error instanceof SpinnerWarnError) {
    stop('warn', error.spinnerText);
  } else {
    stop('fail');
  }

  throw error;
};

const spinnerAction = <T>(action: Promise<T>, text: string): Promise<T> => {
  start(text);

  return action.then(succeed, fail);
};

const spinnerWatchAction = <T>(actions: Array<Promise<T>>, spawnKillHandler: SpawnKillHandler): Promise<T[]> => {
  start('Watch-Mode - exit with "q"');

  const stdin = getStdin();

  stdin.setRawMode(true);
  stdin.setEncoding('utf8');
  stdin.resume();

  stdin.on('data', (key) => {
    // ctrl-c ( end of text )
    if (key.toString() === '\u0003') {
      stdin.setRawMode(false);
      stdin.pause();

      exit();
    }

    if (key.toString() === 'q') {
      stdin.setRawMode(false);
      stdin.pause();

      spawnKillHandler.kill(constants.signals.SIGINT);
    }
  });

  return Promise.all(actions).then(succeed, fail);
};

const spinnerBreak = (): (() => void) => {
  if (isSpinning()) {
    stop('stop');

    return () => instance.start();
  }

  return dummy;
};

export { isSpinning, spinnerAction, spinnerWatchAction, spinnerBreak };
