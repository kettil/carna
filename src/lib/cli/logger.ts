import { isArray, dummy } from '@kettil/tool-lib';
import chalk from 'chalk';
import { spinnerBreak, isSpinning } from './spinner';

type Type = 'stderr' | 'stdout';

type Level = keyof typeof levels;

type Log = (msg: string[] | string) => void;

export type Logger = Record<Level, Log>;

const levels = {
  log: (msg: string) => msg,
  error: chalk.red,
  info: chalk.green,
  debug: chalk.blue,
};

const maxLevelLength = Object.keys(levels).reduce((p, c) => Math.max(p, c.length), 0);

const getMessageLevel = (level: Level) => levels[level](level.toUpperCase().padEnd(maxLevelLength));

const write = (type: Type, message: string) => {
  process[type].write(`${message.trimEnd()}\n`);
};

const logging =
  (level: string, type: Type): Log =>
    (msg) => {
      const resume = spinnerBreak();
      const messages = isArray(msg) ? msg : msg.replace(/\n$/, '').split('\n');

      messages.forEach((m) => {
        write(type, `[${new Date().toISOString()}] ${level}: ${m}`);
      });

      resume();
    };

export const log: Log = (msg) => {
  const indent = isSpinning() ? '  ' : '';
  const resume = spinnerBreak();
  const messages = isArray(msg) ? msg : msg.trimEnd().split('\n');

  write('stdout', messages.map((v) => indent + v).join('\n'));

  resume();
};

const logger = ({ verbose, vvv }: { verbose: boolean; vvv: boolean }): Logger => {
  const error = logging(getMessageLevel('error'), 'stderr');
  const info = logging(getMessageLevel('info'), 'stdout');
  const debug = logging(getMessageLevel('debug'), 'stdout');

  if (vvv) {
    return { log, error, info, debug };
  }

  if (verbose) {
    return { log, error, info, debug: dummy };
  }

  return { log, error: dummy, info: dummy, debug: dummy };
};

export default logger;
