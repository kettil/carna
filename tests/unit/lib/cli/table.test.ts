import { table } from '../../../../src/lib/cli/table';

describe('table', () => {
  test('it should works', () => {
    const data = [
      ['foo', '4'],
      ['bar', '7'],
    ];

    const result = table(data);

    expect(result).toMatchSnapshot();
  });

  test('it should throw an error by inconsistent records', () => {
    const data = [
      ['foo', '4', 'error'],
      ['bar', '7'],
    ];

    const callbackWithException = () => table(data);

    expect(callbackWithException).toThrow('The table rows are of different lengths');
  });
});
