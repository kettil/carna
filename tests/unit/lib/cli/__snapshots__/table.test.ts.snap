// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`table it should works 1`] = `
"┌─────┬───┐
│ foo │ 4 │
├─────┼───┤
│ bar │ 7 │
└─────┴───┘"
`;
